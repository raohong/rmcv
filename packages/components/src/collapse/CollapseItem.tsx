import React, { useRef } from 'react';
import classNames from 'classnames';
import { animated, useSpring } from '@react-spring/web';
import { ArrowDown } from '@rmc-vant/icons';
import { Cell } from '../cell';
import { useConfigContext } from '../config-provider';
import type { CollapseItemProps } from './type';
import { useIsomorphicLayoutEffect } from '../_hooks';

export const COLLAPSEITEM_SYMBOL = Symbol('collapse-item');

const getHeight = (node: HTMLElement | null) => {
  if (!node) {
    return 0;
  }

  const raw = getComputedStyle(node).height;

  node.style.height = 'auto';

  const targetHeight = node.offsetHeight;

  node.style.height = raw;

  return targetHeight;
};

const CollapseItem = React.forwardRef<HTMLDivElement, CollapseItemProps>(
  (
    {
      icon,
      size,
      title,
      value,
      border,
      labelClassName,
      titleClassName,
      valueClassName,
      className,
      disabled,
      children,
      isLink = true,
      toggle,
      collapsed,
      itemKey,
      ...rest
    },
    ref,
  ) => {
    const { getPrefixCls } = useConfigContext();
    const nodeRef = useRef<HTMLDivElement | null>(null);
    const cancelRef = useRef<() => void>();

    const baseCls = getPrefixCls('collapse-item');
    const [{ progress, height }] = useSpring(
      {
        from: {
          height: 0,
          progress: 0,
        },
        to: async (next, cancel) => {
          cancelRef.current = cancel;

          await next({
            height: collapsed ? 0 : getHeight(nodeRef.current),
            progress: collapsed ? 0 : 1,
          });
        },
        config: {
          friction: 36,
          tension: 300,
          velocity: 0.01,
        },
      },
      [collapsed],
    );

    useIsomorphicLayoutEffect(() => {
      return () => {
        cancelRef.current?.();
      };
    }, []);

    const handleToggle: React.MouseEventHandler = () => {
      if (!disabled) {
        toggle?.(itemKey);
      }
    };

    return (
      <div
        ref={ref}
        className={classNames(
          baseCls,
          {
            [`${baseCls}-disabled`]: disabled,
            [`${baseCls}-border`]: border,
          },
          className,
        )}
        aria-expanded={!collapsed}
        {...rest}
      >
        <Cell
          icon={icon}
          size={size}
          title={title}
          value={value}
          className={`${baseCls}-header`}
          labelClassName={classNames(`${baseCls}-label`, labelClassName)}
          valueClassName={classNames(`${baseCls}-value`, valueClassName)}
          titleClassName={classNames(`${baseCls}-title`, titleClassName)}
          isLink={isLink}
          onClick={handleToggle}
          tabIndex={disabled ? -1 : 0}
          role="button"
          aria-disabled={disabled}
          rightIcon={
            <animated.div
              style={{
                rotate: progress.to([0, 1], [0, -180]),
                lineHeight: 0,
              }}
              className={`${baseCls}-right-icon`}
            >
              <ArrowDown />
            </animated.div>
          }
          border={border ? !collapsed : true}
        />
        <animated.div
          ref={nodeRef}
          className={`${baseCls}-wrapper`}
          style={{
            height,
          }}
        >
          <div className={`${baseCls}-content`}>{children}</div>
        </animated.div>
      </div>
    );
  },
);

// @ts-ignore
(CollapseItem as typeof CollapseItem & Record<symbol, any>)[
  COLLAPSEITEM_SYMBOL
] = true;

export default CollapseItem;
