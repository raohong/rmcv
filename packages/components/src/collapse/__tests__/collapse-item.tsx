import { render } from '@testing-library/react';
import CollapseItem from '../CollapseItem';

it('render correctly', () => {
  const tree = render(<CollapseItem />);

  expect(tree.asFragment()).toMatchSnapshot();
});
