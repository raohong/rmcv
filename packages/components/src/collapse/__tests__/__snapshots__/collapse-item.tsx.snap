// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: inherit;
  border: 0;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
  cursor: pointer;
  line-height: 24px;
  padding: 10px 16px;
  font-size: 14px;
  color: #323233;
  background: #fff;
  border-color: #ebedf0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
}

.emotion-1:not(:last-child)::after {
  content: "";
  position: absolute;
  left: -50%;
  top: -50%;
  bottom: -50%;
  right: -50%;
  border-style: solid;
  border-width: 0;
  -webkit-transform: scale(0.5);
  -moz-transform: scale(0.5);
  -ms-transform: scale(0.5);
  transform: scale(0.5);
  pointer-events: none;
  border-color: #ebedf0;
  border-bottom-width: 1px;
}

.emotion-1:not(:last-child)::after {
  margin-left: calc(2 * 16px);
  margin-right: calc(2 * 16px);
}

.emotion-2 {
  font-size: 16px;
  color: #969799;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 4px;
}

.emotion-3 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-4 {
  display: inline-block;
  color: inherit;
  font-style: normal;
  line-height: 0;
  box-sizing: border-box;
  text-rendering: optimizeLegibility;
  -webkit-font-smooth: antialiased;
}

.emotion-4 * {
  line-height: 1;
}

.emotion-5 {
  box-sizing: border-box;
  padding: 12px 16px;
  font-size: 14px;
  color: #969799;
  background: #fff;
}

<div
    aria-expanded="true"
    class="RVuiCollapseItem-root RVuiCollapseItem-expanded emotion-0"
  >
    <div
      aria-disabled="false"
      class="RVuiTouchable-root RVuiCell-root RVuiCell-sizeNormal emotion-1"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="RVuiCell-rightIcon emotion-2"
      >
        <div
          class="RVuiCollapseItem-expandIcon emotion-3"
          style="transform: rotate(-180deg);"
        >
          <span
            aria-label="Arrow-Down"
            class="RVuiIcon-root emotion-4"
            role="img"
          >
            <svg
              fill="currentColor"
              height="1em"
              version="1.1"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505.952256 751.028224a42.467328 42.467328 0 0 1-20.503552-11.393024L143.52384 397.709312c-16.662528-16.661504-16.662528-43.676672 0-60.3392 16.661504-16.662528 43.676672-16.662528 60.3392 0L515.74784 649.253888 827.582464 337.41824c16.661504-16.662528 43.676672-16.662528 60.3392 0s16.662528 43.677696 0 60.340224L545.9968 739.683328c-10.861568 10.861568-26.120192 14.6432-40.044544 11.34592z"
              />
            </svg>
          </span>
        </div>
      </span>
    </div>
    <div
      style="display: none; height: 0px; overflow: hidden;"
    >
      <div
        class="RVuiCollapseItem-content emotion-5"
      />
    </div>
  </div>
</DocumentFragment>
`;
